lib/maths/CTreeShapFeatureImportance.cc

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include <maths/CTreeShapFeatureImportance.h>
+
+#include <algorithm>
+#include <tuple>
+
+namespace ml {
+namespace maths {
+
+using TRowItr = core::CDataFrame::TRowItr;
+
+std::pair<CTreeShapFeatureImportance::TDoubleVecVec, CTreeShapFeatureImportance::TDoubleVec>
+CTreeShapFeatureImportance::shap(const core::CDataFrame& frame,
+                                 const CDataFrameCategoryEncoder& encoder,
+                                 int numberFeatures) {
+    numberFeatures = (numberFeatures != -1) ? numberFeatures : frame.numberColumns();
+    std::vector<std::size_t> maxDepthVec;
+    maxDepthVec.reserve(m_Trees.size());
+    for (auto& tree : m_Trees) {
+        auto samplesPerNode = this->samplesPerNode(tree, frame, encoder);
+        std::size_t maxDepth = this->updateNodeValues(tree, 0, samplesPerNode, 0);
+        maxDepthVec.push_back(maxDepth);
+        m_SamplesPerNode.emplace_back(std::move(samplesPerNode));
+    }
+
+    auto result = frame.readRows(
+        1, 0, frame.numberRows(),
+        core::bindRetrievableState(
+            [&](auto& state, TRowItr beginRows, TRowItr endRows) {
+                for (auto row = beginRows; row != endRows; ++row) {
+                    auto encodedRow{encoder.encode(*row)};
+                    auto& phiVec = state.first;
+                    auto& phiSum = state.second;
+                    TDoubleVec phi(numberFeatures, 0);
+                    for (int i = 0; i < m_Trees.size(); ++i) {
+                        //                        phi[frame.numberColumns()] += m_Trees[i][0].value();
+                        SPath path(maxDepthVec[i] + 1);
+                        this->shapRecursive(m_Trees[i], m_SamplesPerNode[i],
+                                            encodedRow, phi, path, 0, 1.0, 1.0, -1);
+                        i += 0;
+                    }
+                    for (int j = 0; j < phi.size(); ++j) {
+                        phi[j] /= m_Trees.size();
+                        phiSum[j] += std::fabs(phi[j]);
+                    }
+
+                    phiVec.emplace_back(std::move(phi));
+                }
+            },
+            std::make_pair(TDoubleVecVec(), TDoubleVec(numberFeatures, 0))));
```

Comment:
We should `reserve` this vector. However, I'm a bit uncomfortable about writing all these values to a vector of vectors and then writing a subset of them back to the data frame. This particularly breaks matters if the frame is stored off heap for size constraints. I feel like a better strategy is to resize the data frame first and write all the values directly into the data frame storage. When we come to write results we can simply write out the top n.

Reply:
I fixed it by writing directly to the data frame.